src/log.rs

```diff
@@ -1,47 +1,152 @@
+use std::fmt::Write as _;
+use std::io::stderr;
+use std::io::stdout;
+use std::io::Write as _;
 use std::ptr;
 
-extern "C" {
-    fn vfprintf(_: *mut libc::FILE, _: *const c_char, _: ::core::ffi::VaList) -> c_int;
+pub type Dav1dLoggerCallback = unsafe extern "C" fn(
+    // The above `cookie` field.
+    cookie: *mut c_void,
+    // A `printf`-style format specifier.
+    fmt: *const c_char,
+    // `printf`-style variadic args.
+    args: ...
+);
+
+#[derive(Clone)]
+#[repr(C)]
+pub struct Dav1dLogger {
+    /// A cookie that's passed as the first argument to the callback below.
+    cookie: *mut c_void,
+    /// A `printf`-style function except for an extra first argument that will always be the above `cookie`.
+    callback: Option<Dav1dLoggerCallback>,
 }
 
-#[cold]
-pub unsafe extern "C" fn rav1d_log_default_callback(
-    _cookie: *mut c_void,
-    format: *const c_char,
-    mut ap: ::core::ffi::VaList,
-) {
-    vfprintf(stderr, format, ap.as_va_list());
+impl Dav1dLogger {
+    /// # Safety
+    ///
+    /// `callback`, if non-[`None`]/`NULL` must be safe to call when:
+    /// * the first argument is `cookie`
+    /// * the rest of the arguments would be safe to call `printf` with
+    pub const unsafe fn new(cookie: *mut c_void, callback: Option<Dav1dLoggerCallback>) -> Self {
+        Self { cookie, callback }
+    }
 }
 
-impl Default for Rav1dLogger {
+impl Default for Dav1dLogger {
     fn default() -> Self {
-        Self {
-            cookie: ptr::null_mut(),
-            callback: Some(rav1d_log_default_callback),
+        // Safety: `callback` is [`None`].
+        unsafe { Self::new(ptr::null_mut(), None) }
+    }
+}
+
+/// A [`Dav1dLogger`] from C that's not
+/// just a [`Rav1dLogger`] converted to a [`Dav1dLogger`].
+#[derive(Clone)]
+pub(crate) struct OnlyDav1dLogger {
+    cookie: *mut c_void,
+    callback: Dav1dLoggerCallback,
+}
+
+impl fmt::Write for OnlyDav1dLogger {
+    fn write_str(&mut self, s: &str) -> fmt::Result {
+        // `s` doesn't have a terminating nul-byte,
+        // and it may have internal nul-bytes,
+        // so it's easiest just to print one byte at a time.
+        // This may be slow, but logging can be disabled if it's slow,
+        // or the Rust API can be used instead.
+        let fmt = cstr!("%c");
+        for &byte in s.as_bytes() {
+            // # Safety
+            //
+            // The first argument is `self.cookie`
+            // and the rest are safe to call `printf` with,
+            // as required by [`Self::new`].
+            unsafe {
+                (self.callback)(self.cookie, fmt.as_ptr(), byte as c_uint);
+            }
         }
+        Ok(())
     }
 }
 
-#[cold]
-pub unsafe extern "C" fn rav1d_log(c: *mut Rav1dContext, format: *const c_char, args: ...) {
-    if validate_input!(!c.is_null()).is_err() {
-        return;
+#[derive(Clone, Default)]
+pub(crate) enum Rav1dLogger {
+    Dav1d(OnlyDav1dLogger),
+    Stdout,
+    #[default]
+    Stderr,
+}
```

Comment:
What do Rust clients who want a customized logger do here? Ideally they could just provide some `dyn fmt::Write` sink as a config. Would it be simpler and more flexible to just use a trait object here with inner mutability and call it good?

Reply:
> There will presumably be other loggers desired, but it's easier to leave those for later once we know their desired behaviors better.

Yeah, I was thinking that, but it's a bit tricky because `dyn fmt::Write` needs to be boxed, and for stdout and stderr, the common case, we don't want to impose a `Box`, and those implement `io::Write`, which is `write!`able but a bit different, so it's tricky.  I thought we can save those until later when we know better what exact use case is wanted.

Reply:
Yeah, fair enough. I was debating the same thing when I was looking at this.